lark/grammars/python.lark

```diff
@@ -228,14 +228,15 @@ COMMENT: /#[^\n]*/
 STRING : /[ubf]?r?("(?!"").*?(?<!\\)(\\\\)*?"|'(?!'').*?(?<!\\)(\\\\)*?')/i
 LONG_STRING: /[ubf]?r?(""".*?(?<!\\)(\\\\)*?"""|'''.*?(?<!\\)(\\\\)*?''')/is
 
-DEC_NUMBER: /0|[1-9][\d_]*/i
+DEC_NUMBER: /0|\d(?:_?\d)*/i
 HEX_NUMBER.2: /0x[\da-f]*/i
 OCT_NUMBER.2: /0o[0-7]*/i
-BIN_NUMBER.2 : /0b[0-1]*/i
-FLOAT_NUMBER.2: /((\d+\.[\d_]*|\.[\d_]+)([Ee][-+]?\d+)?|\d+([Ee][-+]?\d+))/
-IMAG_NUMBER.2: /\d+[Jj]/ | FLOAT_NUMBER /[Jj]/
+BIN_NUMBER.2: /0b[0-1]*/i
+FLOAT_NUMBER.2: (DEC_NUMBER "." DEC_NUMBER? | "." DEC_NUMBER) (("E" | "e") ["+" | "-"] DEC_NUMBER)?
+              | DEC_NUMBER (("E" | "e") ["+" | "-"] DEC_NUMBER)?
+COMPLEX_NUMBER.2: (DEC_NUMBER | FLOAT_NUMBER) ("+" | "-") (DEC_NUMBER | FLOAT_NUMBER) ("J" | "j")
```

Comment:
This is also wrong. Python doesn't have a complex number literal, but instead the addition happens at runtime (or better, they are combined in the compiler). Most notably, this syntax doesn't allow a lone `1j` or `1j+1`, which are both valid in python.

Reply:
> Python doesn't have a complex number literal, but instead the addition happens at runtime (or better, they are combined in the compiler)

Yeah, that's interesting, in the ast those are two separate constants, while in the byte-code it's already combined
```python
>>> ast.dump(ast.parse("1+5j").body[0].value)
'BinOp(left=Constant(value=1), op=Add(), right=Constant(value=5j))'
>>> dis("1+5j")
  1           0 LOAD_CONST               0 ((1+5j))
              2 RETURN_VALUE
```

> Most notably, this syntax doesn't allow a lone `1j` or `1j+1`, which are both valid in python.

Yep I forgot about it 😅

Well, the reason why I joined it into `COMPLEX_NUMBER` is the convince, but it seems like we can't apply this rule actually.
I never thought of constructions like this 
```python
>>> -1+2j+3+4-5j
(6-3j)
```
We can try to parse this as a complex number but this probably will not be accurate according to CPython's method of parsing

Reply:
Python has a so called peephole optimizer. After all `1+2` without a `j` is also combined into a single `LOAD_CONST`, despite being two separate entries in the AST. We should mirror that behavior.

Reply:
> We should mirror that behavior.

sure

Reply:
resolved in 8a67efbd5ede2ec2bc76e626bb93dfecd5f16ab9